timemachine/fe/single_topology.py

```diff
@@ -0,0 +1,954 @@
+
+    (ytz): we can simplify the API to just bond_idxs, and infer ring atoms and degree from the
+    connectivity alone without resorting to rdkit.
+
+    Parameters
+    ----------
+    ring_bonds: list of 2-tuple
+        Bonds that are part of a ring system.
+
+    proper_idxs: list of 4-tuple
+        Proper torsion idxs
+
+    proper_params: list of 3-tuple
+        Proper torsion parameters
+
+    mol: Chem.Mol
+        RDKit molecule
+
+    Returns
+    -------
+    set
+        Set of chiral bonds.
+
+    """
+    # a chiral bond is defined as a bond that
+    # 1) has a single proper torsion term with k > 10 kJ/mol, period=2, phase=3.1415
+    # 2) is not part of a ring system.
+    # the reason why 2) is present is because the planar torsions spanning a
+    # ring system are not used to enforce chirality, since if we simply
+    # disabled them, we would *still* get the correct chirality due to steric effects.
+    # consider a benzene devoid of any torsions (proper or improper), or non-bonded
+    # terms, and only angles and bonds are present. The hydrogens would still be correctly placed.
+
+    canonical_ring_bonds = set()
+    for ij in ring_bonds:
+        canonical_ring_bonds.add(dummy.canonicalize_bond(ij))
+
+    planar_bonds_kv = identify_bonds_spanned_by_planar_torsions(proper_idxs, proper_params)
+    canonical_chiral_bonds = set()
+    for k in planar_bonds_kv.keys():
+        if k not in canonical_ring_bonds:
+            src, dst = k
+            src = mol.GetAtomWithIdx(src)
+            dst = mol.GetAtomWithIdx(dst)
+            if src.GetDegree() == 4 or dst.GetDegree() == 4:
+                # if either atom is tetrahedral then it has no pi bonds to conjugate over
+                # exception: weird sulfur groups
+                continue
+            else:
+                canonical_chiral_bonds.add(k)
+
+    return canonical_chiral_bonds
+
+
+def find_chiral_atoms(mol):
+    """
+    Find chiral atoms in a molecule. Chirality is perceived locally, without taking
+    graph-symmetry into account. Eg. even something like CH4 would be considered
+    to be chiral since the hydrogens are distinguishable.
+
+    Returns
+    -------
+    set
+        Set of chiral atoms.
+
+    """
+    mol_geom = geometry.classify_geometry(mol)
+    chiral_atoms = set()
+    for a in mol.GetAtoms():
+        a_idx = a.GetIdx()
+        nbs = a.GetNeighbors()
+        if len(nbs) == 4:
+            chiral_atoms.add(a_idx)
+        elif len(nbs) == 3 and mol_geom[a_idx] == LocalGeometry.G3_PYRAMIDAL:
+            # Trivalent Sulfur/Phosphorus typically are not invertible, nor
+            # are they properly parameterized. So we assert if we detect this case.
+            if a.GetAtomicNum() == 16 or a.GetAtomicNum() == 15:
+                assert 0, "Pyramidal Phosphorous or Sulfur detected."
+        elif len(nbs) > 4:
+            assert 0, "More than four neighbors found"
```

Comment:
Aside from the assertions, this could also be implemented with an atomic primitive smarts pattern, if desired

```suggestion
    query_mol = Chem.MolFromSmarts('[X4:1]')
    chiral_atoms = set(match[0] for match in mol.GetSubstructMatches(query_mol))
```

May be more appropriate to move these assertions into one of the TorsionHandlers or similar, if it's mostly meant to catch chemistries that are not well modeled by the forcefield.

Reply:
agreed! I think SMARTS for these probably makes sense.